enumerable_module.rb

```diff
@@ -0,0 +1,312 @@
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    return enum_for :my_each unless block_given?
+
+    index = 0
+    while index < length
+      is_a?(Hash) ? yield([keys[index], values[index]]) : yield(self[index])
+      index += 1
+    end
+    self
+  end
+
+  def my_each_with_index
+    return enum_for :my_each_with_index unless block_given?
+
+    index = 0
+    while index < length
+      is_a?(Hash) ? yield([keys[index], values[index]], index) : yield(self[index], index)
+      index += 1
+    end
+  end
+
+  def my_select
+    return enum_for :my_select unless block_given?
+
+    arr = is_a?(Hash) ? {} : []
+    if is_a?(Hash)
+      my_each { |key, value| arr[key] = value if yield(key, value) }
+    else
+      my_each { |value| arr << value if yield value }
+    end
+    arr
+  end
+
+  def my_all?(arg = nil)
+    extend Snippet
+    if block_given?
+      return my_select { |a, b| yield(a, b) }.length == length if is_a?(Hash)
+
+      my_select { |a| yield(a) }.length == length
+    elsif !arg.nil?
+      all_snippet(arg)
+    else
+      return my_select { |a, _b| a }.length == length if is_a?(Hash)
+
+      my_select { |a| a }.length == length
+    end
+  end
+
+  def my_any?(arg = nil)
+    extend Snippet
+    if block_given?
+      return my_select { |item, value| yield(item, value) }.length.positive? if is_a?(Hash)
+
+      my_select { |item| yield(item) }.length.positive?
+    elsif !arg.nil?
+      any_snippet(arg)
+    else
+      return my_select { |item, _value| item }.length.positive? if is_a?(Hash)
+
+      my_select { |item| item }.length.positive?
+    end
+  end
+
+  def my_none?(arg = nil)
+    extend Snippet
+    if block_given?
+      return my_select { |item, value| yield(item, value) }.length.zero? if is_a?(Hash)
+
+      my_select { |item| yield(item) }.length.zero?
+    elsif !arg.nil?
+      none_snippet(arg)
+    else
+      return my_select { |item, _value| item }.length.zero? if is_a?(Hash)
+
+      my_select { |item| item }.length.zero?
+    end
+  end
+
+  def my_count(arg = nil)
+    arr = []
+    if block_given?
+      my_each { |i| arr << i if yield i }
+    elsif !arg.nil?
+      my_select { |i| arr << i if i == arg }
+    else
+      arr = self
+    end
+    arr.size
+  end
+
+  def my_map(proc = nil)
+    return enum_for :my_map unless block_given?
+
+    arr = []
+    proc ? my_each { |value| arr << proc.call(value) } : my_each { |value| arr << yield(value) }
+    arr
+  end
+
+  def my_inject(*args)
+    arr = is_a?(Range) ? to_a : self
+    if block_given?
+      n = args[0] ? 0 : 1
+      acc = args[0]
+      acc ||= arr[0]
+      n.upto(arr.length - 1) { |value| acc = yield(acc, arr[value]) }
+    elsif args[0].is_a?(Symbol) || args[0].is_a?(Proc)
+      n = 1
+      acc = arr[0]
+      n.upto(arr.length - 1) { |value| acc = args[0].to_proc.call(acc, arr[value]) }
+    end
+    acc
+  end
```

Comment:
**[REQUIRED]**
* When a symbol is specified combines each element of the collection by applying the symbol as a named method.
```
Failure/Error: expect(range.my_inject(2, :*)).to eq range.inject(2, :*)
     
       expected: 2534507766809448170301050680505397403698136797413376000000000000
            got: nil
     
       (compared using ==)
```